cluster-autoscaler/simulator/basic_cluster_snapshot.go

```diff
@@ -0,0 +1,263 @@
+			if podFilter(pod) && selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (data *internalBasicSnapshotData) Pods() schedulerlisters.PodLister {
+	return (*internalBasicSnapshotDataPodLister)(data)
+}
+
+func (data *internalBasicSnapshotData) NodeInfos() schedulerlisters.NodeInfoLister {
+	return (*internalBasicSnapshotDataNodeLister)(data)
+}
+
+// NewEmptySnapshot initializes a Snapshot struct and returns it.
+func newInternalBasicSnapshotData() *internalBasicSnapshotData {
+	return &internalBasicSnapshotData{
+		nodeInfoMap: make(map[string]*schedulernodeinfo.NodeInfo),
+	}
+}
+
+func (data *internalBasicSnapshotData) clone() *internalBasicSnapshotData {
+	clonedNodeInforMap := make(map[string]*schedulernodeinfo.NodeInfo)
+	for k, v := range data.nodeInfoMap {
+		clonedNodeInforMap[k] = v.Clone()
+	}
+	return &internalBasicSnapshotData{
+		nodeInfoMap: clonedNodeInforMap,
+	}
+}
+
+func (data *internalBasicSnapshotData) addNode(node *apiv1.Node) error {
+	if _, found := data.nodeInfoMap[node.Name]; found {
+		return fmt.Errorf("node %s already in snapshot", node.Name)
+	}
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	err := nodeInfo.SetNode(node)
+	if err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo; %v", err)
+	}
+	data.nodeInfoMap[node.Name] = nodeInfo
+	return nil
+}
+
+func (data *internalBasicSnapshotData) removeNode(nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		return fmt.Errorf("node %s not in snapshot", nodeName)
+	}
+	delete(data.nodeInfoMap, nodeName)
+	return nil
+}
+
+func (data *internalBasicSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		return fmt.Errorf("node %s not in snapshot", nodeName)
+	}
+	data.nodeInfoMap[nodeName].AddPod(pod)
+	return nil
+}
+
+func (data *internalBasicSnapshotData) removePod(namespace string, podName string) error {
+	for _, nodeInfo := range data.nodeInfoMap {
+		for _, pod := range nodeInfo.Pods() {
+			if pod.Namespace == namespace && pod.Name == podName {
+				err := nodeInfo.RemovePod(pod)
+				if err != nil {
+					return fmt.Errorf("cannot remove pod; %v", err)
+				}
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("pod %s/%s not in snapshot", namespace, podName)
+}
+
+func (data *internalBasicSnapshotData) getAllPods() ([]*apiv1.Pod, error) {
+	var pods []*apiv1.Pod
+	for _, nodeInfo := range data.nodeInfoMap {
+		pods = append(pods, nodeInfo.Pods()...)
+	}
+	return pods, nil
+}
+
+func (data *internalBasicSnapshotData) getAllNodes() ([]*apiv1.Node, error) {
+	var nodes []*apiv1.Node
+	for _, nodeInfo := range data.nodeInfoMap {
+		nodes = append(nodes, nodeInfo.Node())
+	}
+	return nodes, nil
+}
+
+// NewBasicClusterSnapshot creates instances of BasicClusterSnapshot.
+func NewBasicClusterSnapshot() *BasicClusterSnapshot {
+	snapshot := &BasicClusterSnapshot{}
+	_ = snapshot.Clear()
```

Comment:
nit: this is equivalent to just setting one field in snapshot struct, but it makes me scroll down just to see if there is something more complex going on..

Reply:
But then - if we add something more complex - we will add it to only one place :) Rejected